### PR TITLE
[token-introspection] Create type package

### PR DESCRIPTION
**Pull request by DanielRose**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Adding new definition:

- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by joawan** [comment]
Looks correct to me

File: types/token-introspection/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+// Definitions by: Daniel Rose <https://github.com/DanielRose>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import { RSA_JWK } from 'pem-jwk';
+import { RequestInit, Response } from 'node-fetch';
+import { JwtPayload } from 'jsonwebtoken';
+
+import customErrors = require('./errors');
+
+declare function fetch(url: string, options: RequestInit): Promise<Response>;
```
Comment (peterblazejewicz) [inline types/token-introspection/index.d.ts:12]:
```suggestion
type  Fetch = (url: string, options: RequestInit) => Promise<Response>;
```
should work same IMO, without declaring fake symbol, which doen't exist in source

File: types/token-introspection/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+         * Defaults to [node-fetch](https://github.com/bitinn/node-fetch), but you can inject [zipkin-instrumentation-fetch](https://www.npmjs.com/package/zipkin-instrumentation-fetch).
+         */
+        fetch?: typeof fetch;
+    }
+
+    interface IntrospectResult extends JwtPayload {
+        active: true | 'true';
+    }
+
+    type IntrospectionFunction = (token: string, tokenTypeHint?: string) => Promise<IntrospectResult>;
```
Comment (peterblazejewicz) [inline types/token-introspection/index.d.ts:86]:
```suggestion
   /** @async */
    type IntrospectionFunction = (token: string, tokenTypeHint?: string) => Promise<IntrospectResult>;
```

Comment (peterblazejewicz) [reply types/token-introspection/index.d.ts:86]: (a func/method is declared with `async` keyword, but declaration file has no notion of `async`, so default to JSDoc tag, so people can recognize async/await option)

File: types/token-introspection/errors/ConfigurationError.d.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import IntrospectionError = require('./IntrospectionError');
+
+declare class ConfigurationError extends IntrospectionError {
+    constructor(message?: string);
```
Comment (peterblazejewicz) [inline types/token-introspection/errors/ConfigurationError.d.ts:4]:
```suggestion
```
```ts
interface ErrorConstructor {
    new(message?: string): Error;
    (message?: string): Error;
}
```

File: types/token-introspection/errors/IntrospectionError.d.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+declare class IntrospectionError extends Error {
+    constructor(message?: string);
```
Comment (peterblazejewicz) [inline types/token-introspection/errors/IntrospectionError.d.ts:2]:
```suggestion
```
ditto

File: types/token-introspection/errors/MalformedTokenError.d.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import IntrospectionError = require('./IntrospectionError');
+
+declare class MalformedTokenError extends IntrospectionError {
+    constructor(message?: string);
```
Comment (peterblazejewicz) [inline types/token-introspection/errors/MalformedTokenError.d.ts:4]:
```suggestion
```
ditto

File: types/token-introspection/errors/NotBeforeError.d.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import TokenNotActiveError = require('./TokenNotActiveError');
+
+declare class NotBeforeError extends TokenNotActiveError {
+    constructor(message?: string);
```
Comment (peterblazejewicz) [inline types/token-introspection/errors/NotBeforeError.d.ts:4]:
```suggestion
```
ditto

**Comment by peterblazejewicz** [comment]
@DanielRose 
you can have constructors, but all those have same stamp as super class, so in TS that's superfluous code, up to you.
If you remove constructor, you would have to add exclusion (inline one) for empty class/interface exports, if dt-lint complaints

**Comment by DanielRose** [comment]
@peterblazejewicz Thanks! I implemented the suggested changes, including some tests to make sure they work.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@DanielRose thx!

**Comment by DanielRose** [comment]
Ready to merge

**Pull request merged**